### PR TITLE
Environment refactor

### PR DESCRIPTION
**Pull request by osherwott**
## Description
- Battle Background Images are now retrieved via their own function
  - Because we only need one image, most of the old Environment code was used for this
- Redo environments to be used purely for game mechanics
  - Evolutions:
    - MagneticField, MossRock, IceRock, and Burmy Cloaks
    - Dusty Bowl in the future
  - Water/Dive Ball
  - Caves (used for Sandy Cloak Burmy now but it's nice to have all caves in one place)
  - Hisui areas (also used for just Burmy but might prove useful in the future since they are canon subdivisions of the region)
- getCurrentEnvironments() has been expanded upon to:
  - Use the const `envs` (previously `env`) as an Environment type, to more easily view errors
  - Return a list of Environments in an array
  - Determine Burmy cloaks and combine temp/gym battle environments
#### Burmy
- Burmy cloaks in the games are determined by the tile the battle takes place on, as opposed to a route. Because of this, Cloak environments are listed very generally, and with some overlap, due to our use of Routes. Most routes and all dungeons have been used to set Cloak environments, with only a handful of routes added in service of a better gameplay experience.
- Cloak environments have been limited to regions were Burmy has appeared and has sufficient game data to refer to. This has been done purely to save time on my part and to avoid setting the expectation of putting needless instances where it might change forms in the future, since our playerbase will have no trouble evolving it in the majority of cases, especially with the new generalized areas.
#### Misc:
- Moved Tanoby Ruins to a Water environment and background image to be more faithful to canon, and the map
- Removed Parfum Palace from battle backgrounds const for cleaner code, since it is not needed

#### Part of battle background expansion mentioned in #4232
- Can add more background images without having to change mechanics like Burmy's evolution methods or where Dive balls work
- Can add the second background from #4559 

#### Can now establish more general environments for game mechanics.
- Environment evolutions to highly specific areas to be more faithful to the canon:
  - Eevee:
    - Glaceon can now evolve on Sinnoh Route 217
    - Leafeon can now evolve on Kalos Route 20
    - Moss and Ice rocks NPCs have been added in the towns to mark these locations. In places were the environment is a route, the rocks are placed in the dungeons we've used up until now, and have therefore been included as moss or ice rock environments.
  - Galarian Yamask:
    - Dusty Bowl can now be a route proper
-  Condenses the evolutions on the wiki

#### Creates potential for more environment uses (need to be discussed first, these are only to be used as examples)
- Unique battle environments for temp or gym battles that affect gameplay, ex. damage modifiers
- Deerling form changes, or other pokemon evolutions

## How Has This Been Tested?
This has been tested by running the game and checking various locations the code would affect. I mainly tested early Kanto by including its routes in environments and changing evolution methods of early pokemon. Blue 1 was used to test temp battle area environments being pushed onto the `envs` array. Route 22, Mt. Moon, and both Blue 1 and Brock were used to test the environments being pushed by the general `if` statements for Plant, Sandy, and two Trash cloak environments respectively.
*Temp battle environments stacking with return town environments is intended.*

EDIT: It is now possible to test each area's environment with `MapHelper.getEnvironments()`

## Types of changes
- New feature (for coders)
- UI improvement

### CODE REVIEW

**Comment by Farboo171** [comment]
Is this step 1 of a solution to #2156 ? With follow-up steps to come?

**Comment by osherwott** [comment]
Revised and ready to review

**Comment by osherwott** [comment]
Yeah, I was thinking about Burmy cloaks. I think it's best to be explicit and call those environments `PlantCloak`, `SandyCloak`, and `TrashCloak` environments, especially with Hisui existing. Our Dusk Ball boosts only dungeon catch rate, so what is a Cave shouldn't be a problem, but you're right in that I should make sure it won't be confusing if we ever do decide to use Cave environments.
This PR has been marked as a Draft for the past few weeks, as I don't feel it is ready to be reviewed yet, but I appreciate your feedback nonetheless!

Also I am interested in that `isCurrentEnvironment` function and what exactly you mean by that... but I've been expanding the use of the `getCurrentEnvironment` function, so I will ask properly when I'm done with that!

**Comment by osherwott** [comment]
> I considered having Dusk Balls only work in caves, to be more game-accurate, but that seems like an unnecessary nerf.

Yes, I don't think we want to nerf them, that's why I opted to not change anything regarding them. 👍 

> I hope I'm not stepping on any toes?

Not at all! In fact, I was planning to ask for more feedback from discord at some point. I've just been pleasantly surprised that you cared enough to help me out without being asked! Thank you so much for caring <3

> `isCurrentEnvironment`

I see, so something that would streamline the environment requirement? I'll definitely look into it, and consult discord when I do. Thanks! 👍

File: src/modules/GameConstants.ts [hunk 1]
```diff
@@ -702,10 +702,144 @@ export const ACHIEVEMENT_DEFEAT_DUNGEON_VALUES = [
...
+    CobaltCoastlands: {
+        [Region.hisui]: new Set([23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 'Veilstone Cape', 'Firespit Island', 'Seaside Hollow', 'Turnback Cave', 'Beachside Camp', 'Coastlands Camp', 'Iscan\'s Cabin', 'Molten Arena']),
+    },
+
+    CoronetHighlands: {
+        [Region.hisui]: new Set([38, 39, 40, 41, 42, 43, 44, 45, 46, 'Ancient Wayward Cave', 'Ancient Quarry', 'Primeval Grotto', 'Clamberclaw Cliffs', 'Celestica Ruins', 'Sacred Plaza', 'Temple of Sinnoh', 'Highlands Camp', 'Mountain Camp', 'Summit Camp', 'Moonview Arena']),
+    },
+
+    AlabasterIcelands: {
+        [Region.hisui]: new Set([47, 48, 49, 50, 51, 52, 53, 54, 'Avalugg\'s Legacy', 'Ice Column Chamber', 'Icepeak Cavern', 'Ancient Snowpoint Temple', 'Ancient Lake Acuity', 'Snowfields Camp', 'Icepeak Camp', 'Pearl Settlement', 'Icepeak Arena', 'Stone Portal']),
```
Comment (Qwertypop04) [inline src/modules/GameConstants.ts:832]: Stone Portal is in Coronet

**Review by Qwertypop04 — APPROVED** [review]
Looks good, thanks for the PR!

**Pull request merged**